profiling/src/pcntl.rs

```diff
@@ -0,0 +1,118 @@
+            if pid == 0 {
+                match PROFILER.lock() {
+                    Ok(mut profiler) => {
+                        let mut new_profiler = None;
+                        swap(&mut *profiler, &mut new_profiler);
+
+                        /* Due to the swap, this has the old profiler.
+                         * Forget about it, it has garbage in it.
+                         */
+                        forget(new_profiler);
+                        locals.profiling_enabled = false;
+
+                        /* When we fully support forking remember:
+                         *  - Reset last_cpu_time and last_wall_time
+                         *  - Reset process_id and runtime-id tags.
+                         */
+                    }
+                    Err(err) => {
+                        error!(
+                            "Forked child failed to acquire profiler lock; a crash is likely: {}",
+                            err
+                        )
+                    }
+                }
+            }
+        }
+        Err(r#type) => {
+            warn!(
+                "Return type of pcntl_*fork* function was unexpected: {}",
+                r#type
+            )
+        }
+    }
+}
+
+unsafe fn handle_pcntl_fork(return_value: *mut zval) {
+    /* This hook is installed prior to knowing if the profiler will be enabled,
+     * so we must guard it with a runtime check.
+     */
+    REQUEST_LOCALS.with(|cell| {
+        let locals = cell.borrow_mut();
+        if locals.profiling_enabled && !return_value.is_null() {
+            /* Safety: we just checked it's not null; if any other invariants
+             * are messed up then we're totally hosed already.
+             */
+            handle_child_fork(locals, &mut *return_value);
+        }
+    });
+}
+
+unsafe extern "C" fn pcntl_fork(execute_data: *mut zend_execute_data, return_value: *mut zval) {
+    // Safety: this function is only called if PCNTL_FORK_HANDLER was set.
+    let handler = PCNTL_FORK_HANDLER.unwrap();
+    handler(execute_data, return_value);
+    handle_pcntl_fork(return_value);
+}
+
+unsafe extern "C" fn pcntl_rfork(execute_data: *mut zend_execute_data, return_value: *mut zval) {
+    // Safety: this function is only called if PCNTL_RFORK_HANDLER was set.
+    let handler = PCNTL_RFORK_HANDLER.unwrap();
+    handler(execute_data, return_value);
+    handle_pcntl_fork(return_value);
+}
+
+unsafe extern "C" fn pcntl_forkx(execute_data: *mut zend_execute_data, return_value: *mut zval) {
+    // Safety: this function is only called if PCNTL_FORKX_HANDLER was set.
+    let handler = PCNTL_FORKX_HANDLER.unwrap();
+    handler(execute_data, return_value);
+    handle_pcntl_fork(return_value);
+}
+
+/* Safety: CStr invariants are met (at least one byte, last byte is null,
+ * no null bytes before that). Note that they require their own unsafe
+ * blocks because they are const, despite being in th
```

Comment:
in th?

Reply:
Removed. I used to define these inside of `startup` since they are only used there, but I had to add `unsafe` around them even though `startup` was already an `unsafe` function, so I was starting to document that.